python_modules/dagster/dagster/core/asset_defs/asset_selection.py

```diff
@@ -0,0 +1,142 @@
+        return Resolver(all_assets).resolve(self)
+
+
+class AllAssetSelection(AssetSelection):
+    pass
+
+
+class AndAssetSelection(AssetSelection):
+    def __init__(self, child_1: AssetSelection, child_2: AssetSelection):
+        self.children = (child_1, child_2)
+
+
+class DownstreamAssetSelection(AssetSelection):
+    def __init__(self, child: AssetSelection, *, depth: Optional[int] = None):
+        self.children = (child,)
+        self.depth = depth
+
+
+class GroupsAssetSelection(AssetSelection):
+    def __init__(self, *children: str):
+        self.children = children
+
+
+class KeysAssetSelection(AssetSelection):
+    def __init__(self, *children: str):
+        self.children = children
+
+
+class OrAssetSelection(AssetSelection):
+    def __init__(self, child_1: AssetSelection, child_2: AssetSelection):
+        self.children = (child_1, child_2)
+
+
+class UpstreamAssetSelection(AssetSelection):
+    def __init__(self, child: AssetSelection, *, depth: Optional[int] = None):
+        self.children = (child,)
+        self.depth = depth
+
+
+# ########################
+# ##### RESOLUTION
+# ########################
+
+
+class Resolver:
+    def __init__(self, all_assets: AssetSet):
+        self.all_assets = all_assets
+        self.asset_dep_graph = generate_asset_dep_graph(list(all_assets))
+        self.all_assets_by_name = generate_asset_name_to_definition_map(all_assets)
+
+    def resolve(self, node: AssetSelection) -> AssetSet:
+        if isinstance(node, AllAssetSelection):
+            return self.all_assets
+        elif isinstance(node, AndAssetSelection):
+            child_1, child_2 = [self.resolve(child) for child in node.children]
+            return child_1 & child_2
+        elif isinstance(node, DownstreamAssetSelection):
+            child = self.resolve(node.children[0])
+            return reduce(
+                operator.or_,
+                [self.gather_connected_assets(asset, "downstream", node.depth) for asset in child],
+            )
+        elif isinstance(node, GroupsAssetSelection):
+            return {
+                a
+                for a in self.all_assets
+                if any(_match_group(a, pattern) for pattern in node.children)
+            }
+        elif isinstance(node, KeysAssetSelection):
+            return {a for a in self.all_assets if any(_match_key(a, key) for key in node.children)}
+        elif isinstance(node, OrAssetSelection):
+            child_1, child_2 = [self.resolve(child) for child in node.children]
+            return child_1 | child_2
+        elif isinstance(node, UpstreamAssetSelection):
+            child = self.resolve(node.children[0])
+            return reduce(
+                operator.or_,
+                [self.gather_connected_assets(asset, "upstream", node.depth) for asset in child],
+            )
+        else:
+            check.failed(f"Unknown node type: {type(node)}")
+
+    def gather_connected_assets(
```

Comment:
If this is only used by `resolve`, its name should start with an underscore.

Reply:
good point, fixed